### PR TITLE
spar update issuer

### PR DESCRIPTION
**Pull request by fisx**
https://github.com/zinfra/backend-issues/issues/929

this took about an hour to hack together, and the only thing that's missing is two functions for lookup users on old issuers and for moving them to the new one.  and tests.  and i did some cleanup on the way.  to be specific:

- changed IdP type alias to carry TeamId and [Isuser].
- change db schema to match new type.
- type checker now found some `IdPConfig Teamid` where the type alias `IdP` should be used.
- change errors: when trying to update issuer to one that already exists, that's always bad, not only if it's in another team!

now i should probably spend another hour implementing the competing idea of keeping two idps, adding cassandra columns to make the old and the new idp reference each other, and allow for concurrent use of both.  right now my gut feeling is that it'll be a lot more work to get right.

UPDATE: I just did them both in one PR, seemed easiest.  :)

### CODE REVIEW

**Comment by fisx** [comment]
things to think about more:

- two devices, one using the old login code, one using the new one?
- are login codes cached?
- when do users get logged out and forced to login again?

File: services/spar/test-integration/Test/Spar/APISpec.hs [hunk 1]
```diff
@@ -638,21 +638,97 @@ specCRUDIdentityProvider = do
...
-    describe "issuer changed to one that already exists in another team" $ do
+    describe "issuer changed to one that already exists in *another* team" $ do
       it "rejects" $ do
         env <- ask
         (owner1, _, (^. idpId) -> idpid1) <- registerTestIdP
         (_, _, _, (IdPMetadataValue _ idpmeta2, _)) <- registerTestIdPWithMeta
         callIdpUpdate' (env ^. teSpar) (Just owner1) idpid1 (IdPMetadataValue (cs $ SAML.encode idpmeta2) undefined)
-          `shouldRespondWith` checkErr (== 400) "idp-used-in-other-team"
-    describe "issuer changed to one that is new" $ do
+          `shouldRespondWith` checkErr (== 400) "idp-issuer-in-use"
```
Comment (arianvp) [inline services/spar/test-integration/Test/Spar/APISpec.hs:647]: Why the change in error message type?

Comment (fisx) [reply services/spar/test-integration/Test/Spar/APISpec.hs:647]: i thought it was an error if the idp issuer is already in use in another team, but that's silly.  doesn't matter where the issuer is already in use, whether other team or not.

File: services/spar/test-integration/Test/Spar/APISpec.hs [hunk 2]
```diff
@@ -638,21 +638,97 @@ specCRUDIdentityProvider = do
...
+    describe "issuer changed to one that already exists in *the same* team" $ do
       it "rejects" $ do
         env <- ask
-        (owner, _, (^. idpId) -> idpid, (IdPMetadataValue _ idpmeta, _)) <- registerTestIdPWithMeta
-        newissuer <- makeIssuer
-        let idpmeta' = idpmeta & edIssuer .~ newissuer
-        callIdpUpdate' (env ^. teSpar) (Just owner) idpid (IdPMetadataValue (cs $ SAML.encode idpmeta') undefined)
-          `shouldRespondWith` checkErr (== 400) "cannot-update-idp-issuer"
+        (owner1, _, (^. idpId) -> idpid1, (IdPMetadataValue _ idpmeta1, _)) <- registerTestIdPWithMeta
```
Comment (arianvp) [inline services/spar/test-integration/Test/Spar/APISpec.hs:651]: This... took me at least a 1 minute to parse in my head haha.

Comment (arianvp) [reply services/spar/test-integration/Test/Spar/APISpec.hs:651]: Personally; I'd like if all these helper functions would use properly named datatypes instead of these nested tuples. This is very hard to read without context.

File: services/spar/test-integration/Test/Spar/APISpec.hs [hunk 3]
```diff
@@ -638,21 +638,97 @@ specCRUDIdentityProvider = do
...
+            (_, authnreq) <- call $ callAuthnReq (env ^. teSpar) (idp ^. SAML.idpId)
+            idpresp <- runSimpleSP $ mkAuthnResponseWithSubj userSubject privkey idp spmeta authnreq True
+            sparresp <- submitAuthnResponse idpresp
+            liftIO $ do
+              statusCode sparresp `shouldBe` 200
+              let bdy = maybe "" (cs @LBS @String) (responseBody sparresp)
+              bdy `shouldContain` "<title>wire:sso:success</title>"
+            either (error . show) (pure . view userRefL) $
+              SAML.parseFromDocument (fromSignedAuthnResponse idpresp)
+      context "impure (overwrite old idp)" $ do
```
Comment (arianvp) [inline services/spar/test-integration/Test/Spar/APISpec.hs:675]: can we call this "update existing idp"  instead of impure (overwrite old idp).  I think this is easier to understand for the reader.  to describe what we're testing in terms of the rest 'semantics'

Comment (arianvp) [reply services/spar/test-integration/Test/Spar/APISpec.hs:675]: in current form it makes the reader ask questions,  (why is it impure? why is creation more pure than updating? etc)

File: services/spar/src/Spar/API.hs [hunk 4]
```diff
@@ -278,31 +279,47 @@ validateNewIdP ::
   (HasCallStack, m ~ Spar) =>
   SAML.IdPMetadata ->
   TeamId ->
+  Maybe SAML.IdPId ->
```
Comment (arianvp) [inline services/spar/src/Spar/API.hs:282]: Could we add a haddock explaining the effect of `replaced` parameter here?

File: services/spar/src/Spar/API.hs [hunk 5]
```diff
@@ -251,17 +251,18 @@ idpDelete zusr idpid = withDebugLog "idpDelete" (const Nothing) $ do
...
-idpCreate :: Maybe UserId -> IdPMetadataInfo -> Spar IdP
-idpCreate zusr (IdPMetadataValue raw xml) = idpCreateXML zusr raw xml
+idpCreate :: Maybe UserId -> IdPMetadataInfo -> Maybe SAML.IdPId -> Spar IdP
+idpCreate zusr (IdPMetadataValue raw xml) midpid = idpCreateXML zusr raw xml midpid
 
 -- | We generate a new UUID for each IdP used as IdPConfig's path, thereby ensuring uniqueness.
-idpCreateXML :: Maybe UserId -> Text -> SAML.IdPMetadata -> Spar IdP
-idpCreateXML zusr raw idpmeta = withDebugLog "idpCreate" (Just . show . (^. SAML.idpId)) $ do
+idpCreateXML :: Maybe UserId -> Text -> SAML.IdPMetadata -> Maybe SAML.IdPId -> Spar IdP
+idpCreateXML zusr raw idpmeta mReplaced = withDebugLog "idpCreate" (Just . show . (^. SAML.idpId)) $ do
```
Comment (arianvp) [inline services/spar/src/Spar/API.hs:259]: `replaced` is confusing to me.   `replaces` sounds better?   when we create this Idp it _replaces_ the one in the query parameter no?

File: services/spar/src/Spar/Types.hs [hunk 6]
```diff
@@ -69,7 +69,25 @@ setBindCookieValue = BindCookie . cs . setCookieValue . SAML.fromSimpleSetCookie
...
+
+data WireIdP
+  = WireIdP
+      { _wiTeam :: TeamId,
+        -- | list of issuer names that this idp has replaced, most recent first.  this is used
+        -- for finding users that are still stored under the old issuer, see
+        -- 'findUserWithOldIssuer', 'moveUserToNewIssuer'.
+        _wiOldIssuers :: [SAML.Issuer],
+        -- | the issuer that has replaced this one.  this is set iff a new issuer is created
+        -- with the @"repalces"@ query parameter, and it is used to decide whether users not
```
Comment (arianvp) [inline services/spar/src/Spar/Types.hs:82]: typo: replaces

File: services/spar/src/Spar/Data.hs [hunk 7]
```diff
@@ -343,16 +349,27 @@ storeIdPConfig idp = retry x5 . batch $ do
...
   where
     ins :: PrepQuery W IdPConfigRow ()
-    ins = "INSERT INTO idp (idp, issuer, request_uri, public_key, extra_public_keys, team) VALUES (?, ?, ?, ?, ?, ?)"
+    ins = "INSERT INTO idp (idp, issuer, request_uri, public_key, extra_public_keys, team, old_issuers, replaced_by) VALUES (?, ?, ?, ?, ?, ?, ?, ?)"
     byIssuer :: PrepQuery W (SAML.IdPId, SAML.Issuer) ()
     byIssuer = "INSERT INTO issuer_idp (idp, issuer) VALUES (?, ?)"
     byTeam :: PrepQuery W (SAML.IdPId, TeamId) ()
     byTeam = "INSERT INTO team_idp (idp, team) VALUES (?, ?)"
 
+markReplacedIdP ::
```
Comment (arianvp) [inline services/spar/src/Spar/Data.hs:362]: Can we give this function a name so that it is unambigious which argument is which? or maybe a newtype so we do not mix them up? type are cheap

File: services/spar/src/Spar/API.hs (hunk 5 above)
Comment (arianvp) [reply services/spar/src/Spar/API.hs:259]: or rename the query parameter. Either way works I guess.    or `toReplace` ?

File: services/spar/src/Spar/API.hs [hunk 8]
```diff
@@ -278,31 +279,47 @@ validateNewIdP ::
...
-
-idpUpdateXML :: Maybe UserId -> SAML.IdPId -> Text -> SAML.IdPMetadata -> Spar IdP
-idpUpdateXML zusr idpid raw idpmeta = withDebugLog "idpUpdate" (Just . show . (^. SAML.idpId)) $ do
+-- | FUTUREWORK: 'idpUpdateXML' is only factored out of this function for symmetry with
+-- 'idpCreate', which is not a good reason.  make this one function and pass around
+-- 'IdPMetadataInfo' directly where convenient.
+idpUpdate :: Maybe UserId -> IdPMetadataInfo -> SAML.IdPId -> Spar IdP
+idpUpdate zusr (IdPMetadataValue raw xml) idpid = idpUpdateXML zusr raw xml idpid
+
+-- | The @isPure@ param only has an effect if the issuer changes.  If 'True', a new IdP is
```
Comment (arianvp) [inline services/spar/src/Spar/API.hs:305]: `isPure` is not a parameter anymore. comment is old

File: services/spar/test-integration/Test/Spar/APISpec.hs (hunk 2 above)
Comment (fisx) [reply services/spar/test-integration/Test/Spar/APISpec.hs:651]: Yes.  I think I have a FUTUREWORK about that somewhere.

File: services/spar/test-integration/Test/Spar/APISpec.hs (hunk 3 above)
Comment (fisx) [reply services/spar/test-integration/Test/Spar/APISpec.hs:675]: hum, i just realized that this `context` makes no sense any more, so i removed it...

File: services/spar/src/Spar/API.hs (hunk 8 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs]: 8c0c70b6baa38c1f0f3185b21d34b25b0c0617e6

File: services/spar/src/Spar/Data.hs (hunk 7 above)
Comment (fisx) [reply services/spar/src/Spar/Data.hs:368]: 79a68dbca2272ce2ff536a8ea3be1860be841635

File: services/spar/src/Spar/API.hs (hunk 5 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs]: 3dddc3ebb1395dc43d3d7188390a03d7e3a3a48c

I don't know what that name should be, but at least now I hope it's consistent.

I also found this: 8c0c70b6baa38c1f0f3185b21d34b25b0c0617e6

File: services/spar/src/Spar/API.hs (hunk 4 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs:290]: 173cc64c28b1fc784f239dd7e9c4004451158901

**Comment by fisx** [comment]
Any C* experts who would like to comment on e2dddeb?  @jschaul?

**Comment by fisx** [comment]
> Any C* experts who would like to comment on [e2dddeb](https://github.com/wireapp/wire-server/commit/e2dddeb45a65620e3609be5b116d09d6935021db)? @jschaul?

(I'm happy to give context in a quick phone call tomorrow.)

File: services/spar/src/Spar/API.hs [hunk 9]
```diff
@@ -248,25 +250,49 @@ idpDelete zusr idpid = withDebugLog "idpDelete" (const Nothing) $ do
...
   return NoContent
+  where
+    updateOldIssuers :: IdP -> Spar ()
+    updateOldIssuers _ = pure ()
+    -- we *could* update @idp ^. SAML.idpExtraInfo . wiReplacedBy@ to not keep the idp about
+    -- to be deleted in its old issuers list, but it's tricky to avoid race conditions, and
+    -- there is little to be gained here: we only use old issuers to find users that have not
+    -- been migrated yet, and if an old user points to a deleted idp, it just means that we
+    -- won't find any users to migrate.  still, doesn't hurt mucht to look either.  so we
+    -- leave old issuers dangling for now.
```
Comment (jschaul) [inline services/spar/src/Spar/API.hs:261]: Instead of dangling, could there be an extra field added which would be set here, "marking" these somehow, for a future cleanup?

File: services/spar/src/Spar/API.hs [hunk 10]
```diff
@@ -248,25 +250,49 @@ idpDelete zusr idpid = withDebugLog "idpDelete" (const Nothing) $ do
...
+
+    updateReplacingIdP :: IdP -> Spar ()
+    updateReplacingIdP idp = forM_ (idp ^. SAML.idpExtraInfo . wiOldIssuers) $ \oldIssuer -> do
+      wrapMonadClient $ do
+        iid <- Data.getIdPIdByIssuer oldIssuer
+        mapM_ (Data.clearReplacedBy . Data.Replaced) iid
```
Comment (jschaul) [inline services/spar/src/Spar/API.hs:267]: same comment here as for `setReplacedBy`: This can insert a record into the idp table with only an idp id and no other fields, in certain cases, leading to inconsistent data. Can we guard against that? Perhaps add a comment to this end as well to remind someone in the future refactoring this code and using these queries.

File: services/spar/src/Spar/API.hs (hunk 9 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs:261]: you can easily do the future cleanup by checking for each issuer if there are users *or* the idp entry for it.  if neither, they can be safely removed.

File: services/spar/src/Spar/API.hs (hunk 10 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs:267]: The idp exists in line 266.  There is a possibility that it gets removed between 266 and 267, but that requires two admins working against each other on the same task, *and* a lot of luck.

But I like your suggestions to make the cql code more robust, and I'll look into them!

File: services/spar/schema/src/V9.hs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  ( migration,
+  )
+where
+
+import Cassandra.Schema
+import Imports
+import Text.RawString.QQ
+
+migration :: Migration
+migration = Migration 9 "Keep track of replacing (new) issuers in the idp table" $ do
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/spar/schema/src/V9.hs]: fca9ba816cf824f3efb516fc622b5306ea6c8c4a

File: services/spar/src/Spar/API.hs (hunk 4 above)
Comment (fisx) [reply services/spar/src/Spar/API.hs:307]: I was hoping [this](https://github.com/wireapp/wire-server/commit/173cc64c28b1fc784f239dd7e9c4004451158901) would explain things?

File: services/spar/test-integration/Test/Spar/APISpec.hs (hunk 1 above)
Comment (fisx) [reply services/spar/test-integration/Test/Spar/APISpec.hs:647]: The former label is not in prod yet, and I haven't communicated it to anybody, so it should be fine.

File: services/spar/src/Spar/App.hs [hunk 12]
```diff
@@ -320,17 +340,21 @@ verdictHandlerResultCore bindCky = \case
...
-        (Nothing, Nothing) -> autoprovisionSamlUser userref Nothing ManagedByWire
+        (Nothing, Nothing, Nothing) -> autoprovisionSamlUser userref Nothing ManagedByWire
+        (Nothing, Nothing, Just (oldUserRef, uid)) -> moveUserToNewIssuer oldUserRef userref uid >> pure uid
         -- SSO reauthentication
-        (Nothing, Just uid) -> pure uid
+        (Nothing, Just uid, _) -> pure uid
         -- Bind existing user (non-SSO or SSO) to ssoid
-        (Just uid, Nothing) -> bindUser uid userref
-        (Just uid, Just uid')
+        (Just uid, Nothing, _) -> bindUser uid userref
```
Comment (arianvp) [inline services/spar/src/Spar/App.hs:356]: don't we need to take into account the moved issuer here?

File: services/spar/src/Spar/App.hs [hunk 13]
```diff
@@ -320,17 +340,21 @@ verdictHandlerResultCore bindCky = \case
...
+        (Just uid, Just uid', _)
```
Comment (arianvp) [inline services/spar/src/Spar/App.hs:357]: This case is fine I think

File: services/spar/test-integration/Test/Spar/APISpec.hs [hunk 14]
```diff
@@ -638,21 +638,80 @@ specCRUDIdentityProvider = do
...
-        callIdpUpdate' (env ^. teSpar) (Just owner) idpid (IdPMetadataValue (cs $ SAML.encode idpmeta') undefined)
-          `shouldRespondWith` checkErr (== 400) "cannot-update-idp-issuer"
+        (owner1, _, (^. idpId) -> idpid1, (IdPMetadataValue _ idpmeta1, _)) <- registerTestIdPWithMeta
+        (idpmeta2, _) <- makeTestIdPMetadata
+        _ <- call $ callIdpCreate (env ^. teSpar) (Just owner1) idpmeta2
+        let idpmeta3 = idpmeta1 & edIssuer .~ (idpmeta2 ^. edIssuer)
+        callIdpUpdate' (env ^. teSpar) (Just owner1) idpid1 (IdPMetadataValue (cs $ SAML.encode idpmeta3) undefined)
+          `shouldRespondWith` checkErr (== 400) "idp-issuer-in-use"
+    describe "issuer changed to one that is new" $ do
+      it "updates old idp, updating both issuer and old_issuer" $ do
```
Comment (arianvp) [inline services/spar/test-integration/Test/Spar/APISpec.hs:658]:
```suggestion
      it "updates old idp, updating both issuer and old_issuers" $ do
```

File: services/spar/src/Spar/API.hs [hunk 15]
```diff
@@ -248,25 +250,49 @@ idpDelete zusr idpid = withDebugLog "idpDelete" (const Nothing) $ do
...
+idpCreateXML :: Maybe UserId -> Text -> SAML.IdPMetadata -> Maybe SAML.IdPId -> Spar IdP
+idpCreateXML zusr raw idpmeta mReplaces = withDebugLog "idpCreate" (Just . show . (^. SAML.idpId)) $ do
   teamid <- Brig.getZUsrOwnedTeam zusr
   Galley.assertSSOEnabled teamid
-  idp <- validateNewIdP idpmeta teamid
+  idp <- validateNewIdP idpmeta teamid mReplaces
   wrapMonadClient $ Data.storeIdPRawMetadata (idp ^. SAML.idpId) raw
   SAML.storeIdPConfig idp
+  forM_ mReplaces $ \replaces -> wrapMonadClient $ do
+    Data.setReplacedBy (Data.Replaced replaces) (Data.Replacing (idp ^. SAML.idpId))
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/spar/src/Spar/API.hs:283]: eda57627b39ce42a322357f2e551f76cedcf449a 
5055e61a0903ff8af5326bd4bd2847347c137269

File: services/spar/src/Spar/App.hs (hunk 12 above)
Comment (fisx) [reply services/spar/src/Spar/App.hs]: b9efae06695c19e13188ac24c96e9bf6e6efd1f6
